[SPARK-8313] R Spark packages support

brkyvz opened the pull request:
@shivaram @cafreeman Could you please help me in testing this out? Exposing and running `rPackageBuilder` from inside the shell works, but for some reason, I can't get it to work during Spark Submit. It just starts relaunching Spark Submit.

For testing, you may use the R branch with [sbt-spark-package](https://github.com/databricks/sbt-spark-package). You can call spPackage, and then pass the jar using `--jars`.

brkyvz committed:
R packages support

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36218 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36218/consoleFull) for   PR 7139 at commit [`8810beb`](https://github.com/apache/spark/commit/8810beb215e8b3b646e0ae32794a16dd5da1eb65).

SparkQA commented:
[Test build #36218 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36218/console) for   PR 7139 at commit [`8810beb`](https://github.com/apache/spark/commit/8810beb215e8b3b646e0ae32794a16dd5da1eb65).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

brkyvz committed:
fixed issues

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36294 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36294/consoleFull) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).

shivaram commented:
@brkyvz Thanks for sending out this PR. Its looking good. I had a couple of high level points
1. It might be good to point out in some of the error messages how the JAR should be structured for this to work. I know that it works out of the box with the SBT plugin but it will be good explain this for users who aren't using the plugin

2. Does this also do the package install on all the executors ? Its not important right now with the DataFrame API but some of the work we do in the future will run R code in the executors.

SparkQA commented:
[Test build #36294 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36294/console) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

shivaram commented:
@JoshRosen @shaneknapp -- So in this case the SparkR unit tests failed but the AmpLabJenkins message says Test Passed ? Do you know what could cause this ?

brkyvz commented:
@shivaram Thanks for the feedback. I'll add more error messages regarding structuring. Regarding (1),
right now, this doesn't install anything on the executors, but the procedure should be as simple as running the same command once the executors receive the jars.

shaneknapp commented:
@shivaram -- i'm looking in to the bash monstrosities and seeing if it's somehow missing an exit code...  to this end i'm setting up some test builds on our staging server.  i'll report back once i figure it out.

brkyvz committed:
fix null bug

shaneknapp commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36447 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36447/consoleFull) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).

SparkQA commented:
[Test build #36447 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36447/console) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shaneknapp commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36588/consoleFull) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).

SparkQA commented:
[Test build #36588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36588/console) for   PR 7139 at commit [`0226768`](https://github.com/apache/spark/commit/0226768c1cddce134253b9d817098880f04de0a5).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36604/consoleFull) for   PR 7139 at commit [`bb751ce`](https://github.com/apache/spark/commit/bb751ce029daa90df951bf37843f9e29303ddf64).

SparkQA commented:
[Test build #36604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36604/console) for   PR 7139 at commit [`bb751ce`](https://github.com/apache/spark/commit/bb751ce029daa90df951bf37843f9e29303ddf64).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

brkyvz committed:
added doc

brkyvz committed:
Merge branch 'master' of github.com:apache/spark into r-submit

brkyvz committed:
ready for review

brkyvz commented:
@shivaram @cafreeman I believe this is ready. I added unit, and end to end tests.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36898 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36898/consoleFull) for   PR 7139 at commit [`eff5ba1`](https://github.com/apache/spark/commit/eff5ba15333085adb9f95f1a953cf2c5f506fd2a).

SparkQA commented:
[Test build #36898 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36898/console) for   PR 7139 at commit [`eff5ba1`](https://github.com/apache/spark/commit/eff5ba15333085adb9f95f1a953cf2c5f506fd2a).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

brkyvz committed:
add apache header

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36910/consoleFull) for   PR 7139 at commit [`d867756`](https://github.com/apache/spark/commit/d8677565957f9c1bda03f9bc7cc531684548f9e3).

SparkQA commented:
[Test build #36910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36910/console) for   PR 7139 at commit [`d867756`](https://github.com/apache/spark/commit/d8677565957f9c1bda03f9bc7cc531684548f9e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
Thanks @brkyvz for the update. @sun-rui Could you also take a look at this ?

brkyvz commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36945 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36945/consoleFull) for   PR 7139 at commit [`d867756`](https://github.com/apache/spark/commit/d8677565957f9c1bda03f9bc7cc531684548f9e3).

brkyvz commented:
The test is failing, because Jenkins is not allowing me to install an R package to `$SPARK_HOME/R/lib`. I'll have to update the code somehow. One question I have is, is there a way to disable some tests for people that don't have R? Because we directly use `R CMD INSTALL`, these tests will fail for a random person testing this.

SparkQA commented:
[Test build #36945 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36945/console) for   PR 7139 at commit [`d867756`](https://github.com/apache/spark/commit/d8677565957f9c1bda03f9bc7cc531684548f9e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
So one thing is that we don't need to write a `$SPARK_HOME/R/lib`. If you don't pass in the `-l` argument the package will get installed to the user's home directory which should be writable. 

And regarding the R unit tests, I don't think the SparkR tests get run by maven by `mvn test`. It is explicitly included in our jenkins scripts. One thing we could do is check if R is installed and if not skip the tests ?

cafreeman commented:
@shivaram The weird thing about R CMD INSTALL in this case is that if you don't pass `-l` it'll actually install to the first location on your `.libpaths()` list, which COULD still be `$SPARK_HOME/R/lib` depending on when [this](https://github.com/apache/spark/blob/master/R/pkg/inst/profile/general.R#L20) function gets processed.

If I remember right, there's something that ensures that SparkR is the last library that gets loaded, correct? If that's the case, then R CMD INSTALL would probably default to the standard package install location for R since that's all that would exist on `.libPaths()`.

shivaram commented:
@cafreeman That function only gets executed when we put `shell.R` or `general.R` in `R_PROFILE_USER` (We do this in our spark-submit launchers). If you just run plain R, it shouldn't put SparkR on the path at all. 

FWIW the SparkR being the last package loaded is done at https://github.com/apache/spark/blob/master/R/pkg/inst/profile/shell.R#L25

AmplabJenkins commented:
Merged build triggered.

sun-rui commented:
@brkyvz, could you give more explaination on your usage scenario that this PR is expected to support?

1. This PR introduces a manifest keyword, a hybrid JAR format containing both R code and java classes that the R code may depend on. It feels not so natural, I'd rather like:
   use --jars or --packages to specify JARs purely containing JAVA classes
   introduces new spark-submit flags like --R-src-packages / --R-binary-packages allowing user to specify R packges to be used. No hybrid format is required.

2. This PR installs the R packges only in the local host, which makes it less useful for production cluster environment. For example, for YARN cluster/ Standalone Cluster mode, It is still required that R package to be installed on Driver node (assume DataFrame API). So I would hope support for distributing R source packages or binary packages to Driver and worker nodes (also need to install source packages). Need further discussion on this.

sun-rui commented:
@brkyvz, more comments:)
3. Is it possible to run R specific tests only when -psparkr (sparkr profile) is specified? If sparkr profile is specified when building spark, it is assumed that the user has installed R on the machine.

4. Just FYI, now SparkR supports distributing binary SparkR package to driver and worker nodes. you can take a look at https://issues.apache.org/jira/browse/SPARK-6797. Is it possible to base your work on this?

brkyvz commented:
@sun-rui Thanks for the comments! This PR is basically aiming to support complex Spark Packages, basically packages that have Scala / Java, Python, and/or R code in them. Not just pure R code. Devs will basically write their code in Scala, and provide R APIs. I believe @cafreeman is working on some packages that have this structure.

1. We handle all Maven dependency related information through the `--packages` flag. For example, a package @cafreeman is working on depends on amplab/mlmatrix on the Scala side, and it has R APIs. We added the `--packages` flag to handle situations like this. IMHO I'm against adding more Spark Submit arguments, that would increase complexity. We have Jars that contain Python code in addition to Scala, that is being handled through `--packages` right now.

2. I wanted to keep this PR simple, lay the ground work, and support those cases in follow-up PRs. I can handle those cases in this PR, or follow up like I intentionally planned. Depends on how big a PR you are willing to code review :)

3. I was looking for that exact configuration, thanks! I'll add that.

4. I know, I believe we started working on these in parallel. I can reuse the existing code there.

What do you think? How should I proceed on this PR?

sun-rui commented:
@brkyvz, Thanks for your explanation of this requirement from Devs.

1. Is there any existing JIRA or discussion which you can point me to so that I can learn more on complex Spark package? If this requirement is agreed, we need documentation and or maybe some small tools that can help user to create/examine such packages?

2. This is fine:)

brkyvz commented:
@sun-rui Unfortunately there is no concrete JIRA regarding a discussion. Regarding extra tools, and documentation, there is documentation [here](http://spark-packages.org/artifact-help) and extra tools are:
[sbt-spark-package](https://github.com/databricks/sbt-spark-package) and [spark-package command line tool](https://github.com/databricks/spark-package-cmd-tool).

Right now, the R branch in [sbt-spark-package](https://github.com/databricks/sbt-spark-package) helps you build complex packages with R. I'm going to add support to the spark-package command line tool after we merge this in.

brkyvz committed:
Merge branch 'master' of github.com:apache/spark into r-submit

shivaram commented:
@brkyvz @sun-rui @cafreeman My take is that we should add whatever support we need from the Spark framework in this JIRA and we can then improve it as we go along. I think the important thing here is to have (a) an example which shows how to write a simple R package that only uses R code (b) a more complex example package that uses both R and Scala (I think @cafreeman already has this at https://github.com/cafreeman/sparkGLM ?).

Looks like the PR doesn't merge right now -- @brkyvz can you bring this up to date with master ?

brkyvz committed:
merged master

brkyvz committed:
addressed comments

brkyvz commented:
@sun-rui @shivaram Could you please take a look? Since the codefreeze is right around the corner, I would like this to get in ASAP. In order to make sure that the R package always exists on the driver, I made it a part of SparkContext initialization. Don't know how good of an idea it is, but would love to hear your feedback.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39095/consoleFull) for   PR 7139 at commit [`9778e03`](https://github.com/apache/spark/commit/9778e03916c1f605c4003a935b15135e02567a79).

SparkQA commented:
[Test build #39095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39095/console) for   PR 7139 at commit [`9778e03`](https://github.com/apache/spark/commit/9778e03916c1f605c4003a935b15135e02567a79).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

brkyvz committed:
add assumes for tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

brkyvz committed:
add println ignores

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39100/consoleFull) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39100 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39100/console) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

brkyvz commented:
jenkins retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #171 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/171/consoleFull) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39122/consoleFull) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RBackend.scala:
```diff
@@ -28,6 +28,7 @@ import io.netty.channel.socket.SocketChannel
 import io.netty.channel.socket.nio.NioServerSocketChannel
 import io.netty.handler.codec.LengthFieldBasedFrameDecoder
 import io.netty.handler.codec.bytes.{ByteArrayDecoder, ByteArrayEncoder}
+import org.apache.spark.deploy.RPackageUtils
```
Unused import ?

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RUtils.scala:
```diff
@@ -62,4 +64,7 @@ private[spark] object RUtils {
   }
+
+  /** Check if R is installed before running tests that use R commands. */
+  def isRInstalled: Boolean = Seq("R", "--version").! == 0
```
One minor thing is that this might not work in Windows setups as I think R is shipped as `R.exe` in that case. Any ideas if we actually run tests on Windows as a part of our release process?

brkyvz replied:
I don't think we run windows, therefore it's been a problem in the past.

shivaram commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:
```diff
@@ -22,6 +22,9 @@ import java.net.URL
 import java.util.concurrent.TimeUnit
 
+import org.apache.spark.deploy.IvyTestUtils
+import org.apache.spark.deploy.SparkSubmitUtils.MavenCoordinate
```
unused imports ?

shivaram commented:
Thanks @brkyvz for the update. I had some minor style comments but I don't quite follow why we had to move the initialization to SparkContext vs. having it in the SparkSubmit code (as it was before) ?

brkyvz commented:
@shivaram, based on @sun-rui's comments: 
Spark Submit runs on the host, not necessarily the driver. The package has to be built on the driver, so that it can be really used by the driver. This is important for running in cluster mode. The Spark Submit solution would have worked for client mode only.

SparkQA commented:
[Test build #171 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/171/console) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39122/console) for   PR 7139 at commit [`e6bf7b0`](https://github.com/apache/spark/commit/e6bf7b0b3cb48a7c7ee99da27e2ac034e55d245a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
Ah I see - so this affects things like the YARN cluster mode where the spark-submit script and the driver are run on different machines ? How does this work out for Python packages ? I'm just asking these questions as making a change to SparkContext initialization is pretty intrusive and I'm seeing if we can avoid that

brkyvz commented:
Yes, I agree. I want to avoid it as well. Seems more like a hack than a real solution. The problem is I don't know the Yarn code path too well. 

For Python, we don't need to build anything. Once they're in the `sys.path` inside a zip file (the jar in our case), everything works out.

shivaram commented:
I see - so the right thing to do is to not build the package on every node which needs this (driver / executor) but to build it once in the spark-submit node and then carry the binary package along with the job in `--files` or `--archives` etc. AFAIK this is how @sun-rui got the YARN cluster mode to work (You can see the zip file of the SparkR package being created at [1])

Can we do something similar here ?

[1] https://github.com/apache/spark/blob/27ae851ce16082775ffbcb5b8fc6bdbe65dc70fc/R/install-dev.sh#L47

brkyvz commented:
Hahha, so we're going to unzip first, build, then zip again. Hmm, if I understand the shell script correctly, it will zip all R packages in the `lib` folder, `SparkR`, and whatever was built. If that zip procedure is performed inside Spark Submit, instead of the launcher, then it might work. I don't know if what I do will break it though, because there are no tests.

shivaram commented:
Well we are unzipping the *source*, building a *binary* and then zipping the *binary*. so its not redundant work :).  The script right now only zips the `lib/SparkR` into `sparkr.zip` i.e. just the binary for the SparkR package. 

Regarding test cases I think the changes here shouldn't affect base SparkR from what I see. But I can help with testing this in YARN mode if that'll help

brkyvz commented:
Can we have multiple R packages in one zip? Should we just zip the whole R/lib directory? Otherwise, it's hard to figure out which library was installed, and users may expect to have the lib available.

shivaram commented:
I think we can have multiple R packages in one zip. If I remember correctly we just unzip the files in every executor (or spark-submit does this) and we include that path into the `libPaths` in R.  But to keep the changes separate we could build two zips -- one containing just the base SparkR and the other containing all the addon packages. If it'll make the code simpler we could put the add-on packages in `lib/spark-packages` or something like that as well.

brkyvz commented:
If we put them in `lib/spark-packages`, can R auto discover them? The `R CMD INSTALL` fails, because they can't find SparkR, if I put them in /lib/spark-packages. How bout I just zip everything else than SparkR, since the name of SparkR isn't going to change.

shivaram commented:
Yep that sounds good to me. BTW the way R finds packages is using the `.libPaths` contents. We explicitly set that to include `lib/` which is why it can find `SparkR`.

brkyvz committed:
added zipping of all libs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

brkyvz commented:
@shivaram moved things back, added zipping of packages. This is pretty much set, but not fool proof. YARN support still might need some extra things, and it might break for Windows, but I would like to get this in before the code freeze, so that the interface is there, and address those issues early in the QA period. What do you think?

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
@brkyvz I'll take another pass later today and try this on YARN etc. We should then be able to get it merged hopefully sometime this weekend.

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39277/consoleFull) for   PR 7139 at commit [`ac45527`](https://github.com/apache/spark/commit/ac45527acd6209f9a7de54802eea89b41b7d6692).

brkyvz commented:
By the way, it might fail tests, because it will try to build the packages next to SparkR, and Jenkins might not allow that. However, I'll address that issue once you make a pass.

SparkQA commented:
[Test build #39277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39277/console) for   PR 7139 at commit [`ac45527`](https://github.com/apache/spark/commit/ac45527acd6209f9a7de54802eea89b41b7d6692).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
@brkyvz I just tried this out on a YARN cluster and I actually ran into an error trying to use the CSV reader package. The command I ran was 
```
./bin/sparkR --packages com.databricks:spark-csv_2.10:1.0.3 --master yarn --num-executors 2 examples/src/main/r/data-manipulation.R  /flights.csv
```

The output I got was 
```
Exception in thread "main" java.lang.IllegalArgumentException: URI is not absolute
        at java.io.File.<init>(File.java:416)
        at org.apache.spark.deploy.RPackageUtils$$anonfun$checkAndBuildRPackage$1.apply(RPackageUtils.scala:168)
        at org.apache.spark.deploy.RPackageUtils$$anonfun$checkAndBuildRPackage$1.apply(RPackageUtils.scala:167)
        at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
        at org.apache.spark.deploy.RPackageUtils$.checkAndBuildRPackage(RPackageUtils.scala:167)
        at org.apache.spark.deploy.SparkSubmit$.prepareSubmitEnvironment(SparkSubmit.scala:302)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:154)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

Any idea what is going on ?

brkyvz commented:
The URI was not properly resolved :( I'll ping you when I update it

brkyvz committed:
fix URI

brkyvz commented:
@shivaram could you please try again now?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

shivaram commented:
I'm on it now. BTW is `sparkGLM` the best external package to try out in YARN ? or is there a smaller demo package that I can use ?

SparkQA commented:
[Test build #39564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39564/consoleFull) for   PR 7139 at commit [`77995df`](https://github.com/apache/spark/commit/77995df31fc378e12408ce2b8ec1b6883720f2c2).

brkyvz commented:
I've been using `sparkGLM` for now. You should be able to test it as:
```
library(sparkGLM)
helloWorldTest("World")
```

brkyvz commented:
I've been passing it through `--jars` though, since it's not available anywhere.

shivaram commented:
@brkyvz I tried to run this with YARN and found that the code to create the `zip` might have some problems. It ends up creating a zip file which is multiple gigabytes and the FS runs out of space etc. Could you point me to where the code to create the sparkr-extras.zip is ?

brkyvz commented on core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  /** Zips all the libraries found with SparkR in the R/lib directory for distribution with Yarn. */
+  private[deploy] def zipRLibraries(dir: File, name: String): File = {
```
This method creates the zip

brkyvz commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -367,9 +376,11 @@ object SparkSubmit {
         printErrorAndExit(s"$SPARKR_PACKAGE_ARCHIVE does not exist for R application in YARN mode.")
       }
       val localURI = Utils.resolveURI(rPackageFile.getAbsolutePath)
+      val rExtras = RPackageUtils.zipRLibraries(new File(rPackagePath.get), SPARKR_EXTRAS_ARCHIVE)
```
this is where you get the folder and zip

brkyvz commented:
@shivaram commented on the lines

SparkQA commented:
[Test build #39564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39564/console) for   PR 7139 at commit [`77995df`](https://github.com/apache/spark/commit/77995df31fc378e12408ce2b8ec1b6883720f2c2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

brkyvz commented:
I believe I may have to add the `.zip` to the fileName filter in the function above `zipRLibraries` as well.

brkyvz committed:
don't zip

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39601/consoleFull) for   PR 7139 at commit [`3a1be7d`](https://github.com/apache/spark/commit/3a1be7dbc3f91b6591beb81dae2b3f2135e8417f).

SparkQA commented:
[Test build #39601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39601/console) for   PR 7139 at commit [`3a1be7d`](https://github.com/apache/spark/commit/3a1be7dbc3f91b6591beb81dae2b3f2135e8417f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

brkyvz committed:
added zipping test

brkyvz committed:
merged master

AmplabJenkins commented:
Merged build triggered.

brkyvz commented:
@shivaram @cafreeman This should work even on Yarn Cluster mode now! I'm pretty sure this version will fail Jenkins on one test  (SparkSubmit end to end R package test), because Jenkins will probably not allow me install the library. However the code should be ready, could you please take a look and comment?

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39617 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39617/consoleFull) for   PR 7139 at commit [`4258ffe`](https://github.com/apache/spark/commit/4258ffe9633aebdaa28a331a7aab9a286137ae58).

shivaram commented on R/install-dev.sh:
```diff
@@ -44,6 +44,5 @@ R CMD INSTALL --library=$LIB_DIR $FWDIR/pkg/
 
 # Zip the SparkR package so that it can be distributed to worker nodes on YARN
```
You can also remove this comment and the `cd` I guess ?

shivaram commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -46,7 +46,7 @@ import org.apache.mesos.MesosNativeLibrary
 import org.apache.spark.annotation.{DeveloperApi, Experimental}
 import org.apache.spark.broadcast.Broadcast
-import org.apache.spark.deploy.{LocalSparkCluster, SparkHadoopUtil}
+import org.apache.spark.deploy.{RPackageUtils, LocalSparkCluster, SparkHadoopUtil}
```
unused import now

shivaram commented on core/src/main/scala/org/apache/spark/deploy/RPackageUtils.scala:
```diff
@@ -0,0 +1,233 @@
+    }
+  }
+
+  private def listFilesRecursively(dir: File): Set[File] = {
```
Lets add a comment above this saying this lists all files without `.zip` extension ? Actually can you make that an argument to this function ? something like `excludePattern` would be fine

shivaram commented:
Thanks @brkyvz for the update. I did one pass over the code and mostly had minor comments. I think the idea of building a zip file at the end just before we launch is pretty good. BTW there is some code to create the zip in Windows at https://github.com/apache/spark/blob/3b0e44490aebfba30afc147e4a34a63439d985c6/R/install-dev.bat#L29 that can also be cleaned up now. 

BTW I'll also try to test this on Windows and see how far I get. My guess is that if we can make the name of the R binary configurable we should be able to get this working on windows, but we can do that in a follow up PR too

SparkQA commented:
[Test build #39617 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39617/console) for   PR 7139 at commit [`4258ffe`](https://github.com/apache/spark/commit/4258ffe9633aebdaa28a331a7aab9a286137ae58).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
@brkyvz The jenkins failure message seems to be 
```
ERROR: cannot cd to directory '/home/jenkins/workspace/SparkPullRequestBuilder@2/R/lib
```
from https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39617/artifact/core/target/unit-tests.log

Any ideas why we can't use that directory ? Are the directories read only by default or something like that ?

brkyvz commented:
I guess the folder is read-only. Which makes sense, because it is part of the test infrastructure (what if someone added a test that ran `rm -rf`). I don't know of a work around :( Another possibility is that the folder doesn't exist. There is no `R/lib`. Let me try a workaround for the second option. If it's the first, I can't think of any options.

brkyvz commented:
If `R/lib` doesn't exist though, that means that `SparkR` doesn't exist either, which means that the test will fail anyway :/ Maybe I could just remove that test? Or add a SparkR test instead? Is there any that use spark submit?

brkyvz committed:
addressed comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

brkyvz commented:
Hi @shivaram. Addressed your comments. Regarding the failing test... I disabled it for now. Maybe a better solution will be to include a test jar similar to what you already have for `test_includeJAR.R` and run the test on the Spark R side instead of Spark Submit.

SparkQA commented:
[Test build #39666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39666/consoleFull) for   PR 7139 at commit [`6603d0d`](https://github.com/apache/spark/commit/6603d0dcc80b5759f357194f5d102cac51b9be0d).

SparkQA commented:
[Test build #39666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39666/console) for   PR 7139 at commit [`6603d0d`](https://github.com/apache/spark/commit/6603d0dcc80b5759f357194f5d102cac51b9be0d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
@brkyvz I think I figured out the problem with Jenkins -- Since Jenkins uses SBT it doesn't build the SparkR package at the beginning along with the rest of the artifacts and only builds SparkR while running SparkR unit tests. So what ends up happening is that R/lib doesn't exist when the core unit tests are running. 

I think the right thing to do here is to get Jenkins to build SparkR along with other components. @JoshRosen and @yu-iskw were discussing a similar issue in #7883. For now can we just comment out the test and open a JIRA to un-comment it once we fix this ?

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RUtils.scala:
```diff
@@ -19,6 +19,8 @@ package org.apache.spark.api.r
 
 import java.io.File
 
+import scala.collection.JavaConversions._
```
Just checking -- do we need this import ?

brkyvz replied:
for the `Process Builder` below. It takes a Java Collection

brkyvz committed:
removed unused imports

brkyvz committed:
remove unused imports 2

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

brkyvz commented:
@shivaram removed the unused imports. Created a JIRA ([SPARK-9603](https://issues.apache.org/jira/browse/SPARK-9603)) for re-enabling the test, and added it as a TODO to the test

shivaram commented:
Thanks @brkyvz -- Changes LGTM. Can you check if @andrewor14 wants to take another look at the SparkSubmit changes ?

SparkQA commented:
[Test build #39729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39729/consoleFull) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).

brkyvz commented:
jenkins retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #212 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/212/consoleFull) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39738 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39738/consoleFull) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).

SparkQA commented:
[Test build #39729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39729/console) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39747/consoleFull) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).

SparkQA commented:
[Test build #212 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/212/console) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39738 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39738/console) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39747/console) for   PR 7139 at commit [`0de384f`](https://github.com/apache/spark/commit/0de384f33b92252c7e0a9e187a6c83c7e96d4e06).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

pwendell commented:
I took a look at the spark submit changes and they LGTM.

brkyvz commented:
thanks @pwendell, @shivaram can you merge this please?

pwendell commented:
@shivaram maybe you can merge? I looked at the spark submit stuff but overall it was a very small part of the changes.

shivaram commented:
Yep - I'm out now, but will get back to my computer and merge

shivaram commented:
BTW @brkyvz can you merge the R changes in the https://github.com/databricks/sbt-spark-package as well ?

brkyvz commented:
thanks @shivaram. Will do, I'll add the suggested format to the spark-package command line tool as well.
